pkg/ddc/juicefs/metadata.go

```diff
@@ -0,0 +1,213 @@
+	if err != nil {
+		should = false
+		return should, err
+	}
+
+	//todo(xuzhihao): option to enable/disable automatic metadata sync
+	//todo: periodical metadata sync
+	if dataset.Status.UfsTotal != "" && dataset.Status.UfsTotal != METADATA_SYNC_NOT_DONE_MSG {
+		j.Log.V(1).Info("dataset ufs is ready",
+			"dataset name", dataset.Name,
+			"dataset namespace", dataset.Namespace,
+			"ufstotal", dataset.Status.UfsTotal)
+		should = false
+		return should, nil
+	}
+	should = true
+	return should, nil
+}
+
+// syncMetadataInternal does the actual work of metadata sync
+// At any time, there is at most one goroutine working on metadata sync. First call to
+// this function will start a goroutine including the following two steps:
+//   1. load metadata
+//   2. get total size of UFSs
+// Any following calls to this function will try to get result of the working goroutine with a timeout, which
+// ensures the function won't block the following Sync operations(e.g. CheckAndUpdateRuntimeStatus) for a long time.
+func (j *JuiceFSEngine) syncMetadataInternal() (err error) {
+	if j.MetadataSyncDoneCh != nil {
+		// Either get result from channel or timeout
+		select {
+		case result := <-j.MetadataSyncDoneCh:
+			defer func() {
+				j.MetadataSyncDoneCh = nil
+			}()
+			j.Log.Info("Get result from MetadataSyncDoneCh", "result", result)
+			if result.Done {
+				j.Log.Info("Metadata sync succeeded", "period", time.Since(result.StartTime))
+				err = retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+					dataset, err := utils.GetDataset(j.Client, j.name, j.namespace)
+					if err != nil {
+						return
+					}
+					datasetToUpdate := dataset.DeepCopy()
+					datasetToUpdate.Status.UfsTotal = result.UfsTotal
+					datasetToUpdate.Status.FileNum = result.FileNum
+					if !reflect.DeepEqual(datasetToUpdate, dataset) {
+						err = j.Client.Status().Update(context.TODO(), datasetToUpdate)
+						if err != nil {
+							return
+						}
+					}
+					return
+				})
+				if err != nil {
+					j.Log.Error(err, "Failed to update UfsTotal and FileNum of the dataset")
+					return err
+				}
+			} else {
+				j.Log.Error(result.Err, "Metadata sync failed")
+				return result.Err
+			}
+		case <-time.After(CHECK_METADATA_SYNC_DONE_TIMEOUT_MILLISEC * time.Millisecond):
+			j.Log.V(1).Info("Metadata sync still in progress")
+		}
+	} else {
+		// Metadata sync haven't started
+		err = retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+			dataset, err := utils.GetDataset(j.Client, j.name, j.namespace)
+			if err != nil {
+				return
+			}
+			datasetToUpdate := dataset.DeepCopy()
+			datasetToUpdate.Status.UfsTotal = METADATA_SYNC_NOT_DONE_MSG
+			datasetToUpdate.Status.FileNum = METADATA_SYNC_NOT_DONE_MSG
+			if !reflect.DeepEqual(dataset, datasetToUpdate) {
+				err = j.Client.Status().Update(context.TODO(), datasetToUpdate)
+				if err != nil {
+					return
+				}
+			}
+			return
+		})
+		if err != nil {
```

Comment:
This error does not return, is normal business logic?

Reply:
Yes, it's update status with "Calculating". Even if it failed here, metadataSync will still go on.